core/corehttp/gateway_handler_unixfs_dir.go

```diff
@@ -220,3 +201,36 @@ func (i *gatewayHandler) serveDirectory(ctx context.Context, w http.ResponseWrit
 func getDirListingEtag(dirCid cid.Cid) string {
 	return `"DirIndex-` + assets.AssetHash + `_CID-` + dirCid.String() + `"`
 }
+
+func (i *gatewayHandler) getUnixFsLinks(ctx context.Context, cid cid.Cid) ([]*format.Link, error) {
+	obj, err := i.api.Dag().Get(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	protoNode, ok := obj.(*merkledag.ProtoNode)
+	if !ok {
+		return obj.Links(), nil
+	}
+
+	fsNode, err := unixfs.FSNodeFromBytes(protoNode.Data())
+	if err != nil {
+		return nil, err
+	}
+
+	if fsNode.Type() == unixfs.THAMTShard {
+		shard, err := hamt.NewHamtFromDag(i.api.Dag(), obj)
+		if err != nil {
+			return nil, err
+		}
+
+		links, err := shard.EnumLinks(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return links, nil
+	} else {
+		return obj.Links(), nil
+	}
+}
```

Comment:
Note for reviewers: couldn't find an API that did this directly for me. If it exists, please let me know.

Reply:
Maybe @Jorropo knows? He looked at unixfs libs recently.

Reply:
`core/coreapi` is supposed to be that package.

Reply:
> `core/coreapi` is supposed to be that package.

Yes, but only `UnixFs.Ls` is there. And it does much more processing, which is much slower.